lib/Email/Stuffer/TestLinks.pm

```diff
@@ -22,55 +26,74 @@ Email::Stuffer>send_or_die()
 
 =head1 DESCRIPTION
 
-When this module is included in a test, it parses HTML links (<a href="xyz"...)
-in every email sent through Email::Stuffer->send_or_die(). Each URI must get a
-successful response code (200 range) and the returned pagetitle must not contain
-'error' or 'not found'.
+When this module is included in a test, it parses http links (<a href="xyz">...</a>)
+and image links (<img src="xyz">) in every email sent through Email::Stuffer->send_or_die().
+Each URI must be  get a successful response code (200 range).
+Page title must not contain 'error' or 'not found' for text/html content.
+Image links must return an image content type.
 
 =cut
 
 install_modifier 'Email::Stuffer', after => send_or_die => sub {
 
     my $self = shift;
-    my $ua = Mojo::UserAgent->new(max_redirects => 10, connect_timeout => 5);
 
     my %urls;
     $self->email->walk_parts(
         sub {
             my ($part) = @_;
             return unless ($part->content_type && $part->content_type =~ /text\/html/i);
             my $dom = Mojo::DOM->new($part->body);
-            my $links = $dom->find('a')->map(attr => 'href')->compact;
+            push @{$urls{http}}, $dom->find('a')->map(attr => 'href')->compact->grep(sub {$_ !~ /^mailto:/})->uniq->to_array->@*;
+            push @{$urls{image}}, $dom->find('img')->map(attr => 'src')->compact->uniq->to_array->@*;
+    });
 
-            # Exclude anchors, mailto
-            $urls{$_} = 1 for (grep { !/^mailto:/ } @$links);
-        });
+    my @data = map { my $type = $_; map { [ $type, $_ ] } $urls{$type}->@* } keys %urls;
 
-    for my $url (sort keys %urls) {
+    my $loop = IO::Async::Loop->new();
+    $loop->add( my $http = Net::Async::HTTP->new( max_connections_per_host=>3 ) );
 
-        my $err = '';
+    ( fmap_void {
+       my ($type, $url) = @$_;
 
-        if ($url =~ /^[#\/]/) {
-            $err = "$url is not a valid URL for an email";
-        } else {
-            my $tx = $ua->get($url);
+        my $uri = URI->new($url);
+        unless ( $uri->scheme ) {
+            fail "$type link $url is an invalid uri";
```

Comment:
Unless this module is used solely for testing this doesn't sound like a good place to put `fail`

Reply:
Yes, the idea for this module is you include it in a test, and it will inspect all outgoing emails. It's only supposed to be used in tests.